File: elixir/test/math_operations_test.exs
defmodule CornTest.MathOperationsTest do
  use ExUnit.Case, async: true

  alias CornTest.MathOperations

  describe "add/2" do
    test "adds two positive numbers" do
      assert MathOperations.add(2, 3) == 5
      assert MathOperations.add(40, 60) == 100
    end

    test "adds negative numbers" do
      assert MathOperations.add(-2, -3) == -5
      assert MathOperations.add(5, -5) == 0
    end
  end

  describe "subtract/2" do
    test "subtracts two numbers" do
      assert MathOperations.subtract(5, 3) == 2
      assert MathOperations.subtract(3, 5) == -2
      assert MathOperations.subtract(10, 10) == 0
    end
  end

  describe "multiply/2" do
    test "multiplies two numbers" do
      assert MathOperations.multiply(2, 3) == 6
      assert MathOperations.multiply(-2, 3) == -6
      assert MathOperations.multiply(0, 100) == 0
    end
  end

  describe "divide/2" do
    test "divides two numbers" do
      assert MathOperations.divide(6, 3) == {:ok, 2.0}
      assert MathOperations.divide(5, 2) == {:ok, 2.5}
    end

    test "returns error for division by zero" do
      assert MathOperations.divide(5, 0) == {:error, "Division by zero"}
    end
  end

  describe "power/2" do
    test "calculates power" do
      assert MathOperations.power(2, 3) == 8.0
      assert MathOperations.power(5, 0) == 1.0
    end
  end

  describe "factorial/1" do
    test "calculates factorial" do
      assert MathOperations.factorial(0) == {:ok, 1}
      assert MathOperations.factorial(1) == {:ok, 1}
      assert MathOperations.factorial(5) == {:ok, 120}
    end

    test "returns error for negative numbers" do
      assert MathOperations.factorial(-1) == {:error, "Factorial is not defined for negative numbers"}
    end
  end

  describe "derivative/1" do
    test "calculates derivative of polynomial" do
      assert MathOperations.derivative([3, 2, 5]) == [2, 10]
      assert MathOperations.derivative([7]) == [0.0]
      assert MathOperations.derivative([]) == [0.0]
    end
  end

  describe "pi/0" do
    test "returns first 40 decimals of pi" do
      assert MathOperations.pi() == "3.1415926535897932384626433832795028841971"
    end
  end

  describe "gcd/2" do
    test "calculates greatest common divisor" do
      assert MathOperations.gcd(10, 15) == 5
      assert MathOperations.gcd(48, 18) == 6
      assert MathOperations.gcd(17, 19) == 1
    end

    test "handles negative numbers" do
      assert MathOperations.gcd(-10, 15) == 5
      assert MathOperations.gcd(48, -18) == 6
    end
  end
end

File: dart/test/math_operations_test.dart
import 'package:test/test.dart';
import 'package:corn_test_dart/math_operations.dart';

void main() {
  late MathOperations mathOps;

  setUp(() {
    mathOps = MathOperations();
  });

  group('add', () {
    test('adds positive numbers', () {
      expect(mathOps.add(2, 3), equals(5));
      expect(mathOps.add(40, 60), equals(100));
    });

    test('adds negative numbers', () {
      expect(mathOps.add(-2, -3), equals(-5));
      expect(mathOps.add(5, -5), equals(0));
    });
  });

  group('subtract', () {
    test('subtracts two numbers', () {
      expect(mathOps.subtract(5, 3), equals(2));
      expect(mathOps.subtract(3, 5), equals(-2));
    });
  });

  group('multiply', () {
    test('multiplies two numbers', () {
      expect(mathOps.multiply(2, 3), equals(6));
      expect(mathOps.multiply(-2, 3), equals(-6));
      expect(mathOps.multiply(0, 100), equals(0));
    });
  });

  group('divide', () {
    test('divides two numbers', () {
      expect(mathOps.divide(6, 3), closeTo(2.0, 0.001));
      expect(mathOps.divide(5, 2), closeTo(2.5, 0.001));
    });

    test('throws on division by zero', () {
      expect(() => mathOps.divide(5, 0), throwsArgumentError);
    });
  });

  group('power', () {
    test('calculates power', () {
      expect(mathOps.power(2, 3), closeTo(8.0, 0.001));
      expect(mathOps.power(5, 0), closeTo(1.0, 0.001));
    });
  });

  group('factorial', () {
    test('calculates factorial', () {
      expect(mathOps.factorial(0), equals(1));
      expect(mathOps.factorial(1), equals(1));
      expect(mathOps.factorial(5), equals(120));
    });

    test('throws for negative numbers', () {
      expect(() => mathOps.factorial(-1), throwsArgumentError);
    });
  });

  group('derivative', () {
    test('calculates polynomial derivative', () {
      expect(mathOps.derivative([3, 2, 5]), orderedEquals([2.0, 10.0]));
      expect(mathOps.derivative([7]), orderedEquals([0.0]));
      expect(mathOps.derivative([]), orderedEquals([0.0]));
    });
  });

  group('pi', () {
    test('returns first 40 decimals of pi', () {
      expect(mathOps.pi(), equals('3.1415926535897932384626433832795028841971'));
    });
  });

  group('gcd', () {
    test('calculates greatest common divisor', () {
      expect(mathOps.gcd(10, 15), equals(5));
      expect(mathOps.gcd(48, 18), equals(6));
      expect(mathOps.gcd(17, 19), equals(1));
    });

    test('handles negative numbers', () {
      expect(mathOps.gcd(-10, 15), equals(5));
      expect(mathOps.gcd(48, -18), equals(6));
    });
  });
}
